### PR TITLE
Add initial set of query validation tests #668

### PR DESCRIPTION
**Pull request by dguy**
Adds:
1. parameterized test that reads a json file to build and run streams topologies for queries provided.
2. First set of tests
3. Fixes bugs in TopK and TopKDistinct where the results aren't always sorted
4. Fixes bug in Json deserializer where null values are not returned as null
5. Fixes bug in KudafInitializer where the same value is returned each time it is called which can result in incorrect initialization when using reference types

#668

### CODE REVIEW

**Comment by apurvam** [comment]
Thanks for the patch @dguy .. 

1. If I understand correctly, you have a JSON file defining the 'test spec', which includes the query, inputs, and outputs. 
2. You then use the KSQLEngine to compile the given query to a streams topology. You then pass that topology to the `TopologyTestDriver`, which generates outputs for a given sequence of inputs given that topology. 
3. You then verify that the outputs thus received are what you expect.

If my understanding is correct, this looks great to me. It is a very low cost way of verifying that we don't introduce regressions when we change the engine. 

My biggest question is how we can test the KSQL UDAFs like the various aggregates etc.. These are used extensively by our early adopters, and it would be valuable to have tests for those, but I am not sure if it is possible in the current format.

**Comment by dguy** [comment]
@apurvam yeah we can test the aggregates i.e, `count()`, `sum()` etc. I added a couple more example tests that do just this

**Comment by dguy** [comment]
note: the TopK tests currently look incorrect. However, we need https://github.com/confluentinc/ksql/pull/695 to be merged so i can update and see if there any other issues

**Comment by dguy** [comment]
Also, any suggestions on other tests to add?

**Comment by satybald** [comment]
@dguy I'd also consider tests for TOPK DISTINCT.

**Comment by hjafarpour** [comment]
@dguy this is really great tool for testing. I would also add a few tests that include more complex expressions in select items and WHERE clause. For instance a test for computing average by dividing sum and count.

**Comment by sriramsub** [comment]
This is great! Would help us to build a large set of queries for validation pretty quickly

**Comment by dguy** [comment]
This is ready for review now. We can add more tests in follow up PRs, but i'd like to get this in as it has also fixed some bugs (see notes).

File: ksql-engine/src/main/java/io/confluent/ksql/function/udaf/KudafInitializer.java [hunk 1]
```diff
@@ -21,19 +21,30 @@
...
+  public KudafInitializer(final int nonAggValSize) {
+    this.nonAggValSize = nonAggValSize;
   }
 
   @Override
   public GenericRow apply() {
-    List rowColumns = new ArrayList(initialGenericRowColumns);
-    return new GenericRow(rowColumns);
+    final List<Object> values = IntStream.range(0, nonAggValSize)
+        .mapToObj(value -> null)
```
Comment (apurvam) [inline ksql-engine/src/main/java/io/confluent/ksql/function/udaf/KudafInitializer.java]: Not sure I follow this line. Will `values` just be a list of `null` of size `nonAggValSize`?

File: ksql-engine/src/main/java/io/confluent/ksql/function/udaf/KudafAggregator.java [hunk 2]
```diff
@@ -62,11 +62,11 @@ public GenericRow apply(String s, GenericRow rowValue, GenericRow aggRowValue) {
       GenericRow mergedRow = new GenericRow(columns);
 
       aggValToValColumnMap.forEach((columnIndex, value) -> {
-        if (aggRowOne.getColumns().get(value).toString().length() > 0) {
-          mergedRow.getColumns().set(columnIndex, aggRowOne.getColumns()
+        if (aggRowOne.getColumns().get(value) == null) {
+          mergedRow.getColumns().set(columnIndex, aggRowTwo.getColumns()
```
Comment (apurvam) [inline ksql-engine/src/main/java/io/confluent/ksql/function/udaf/KudafAggregator.java]: The logic seems to have been inverted in this change set. Was that intentional? Was the previous behavior a bug?

File: ksql-engine/src/test/resources/query-validation-tests.json [hunk 3]
```diff
@@ -0,0 +1,340 @@
+{
+  "tests": [
```
Comment (apurvam) [inline ksql-engine/src/test/resources/query-validation-tests.json]: Probably would be good to add a comment that you can have multiple statements per test case, but the last is assumed to be a query and is the only one whose topology is verified.

File: ksql-engine/src/main/java/io/confluent/ksql/function/udaf/KudafAggregator.java (hunk 2 above)
Comment (dguy) [reply ksql-engine/src/main/java/io/confluent/ksql/function/udaf/KudafAggregator.java]: yes it was intentional - previously the default value for the columns was `""` hence the string length check. I've changed the default value to be `null`

File: ksql-engine/src/main/java/io/confluent/ksql/function/udaf/KudafInitializer.java (hunk 1 above)
Comment (dguy) [reply ksql-engine/src/main/java/io/confluent/ksql/function/udaf/KudafInitializer.java]: Yes, it needs to be a list of size `numberOfColumns` of empty/default values. This hasn't really changed, i just moved the code here as the list needs to be new instance each time this is called so values are re-initialized, i.e., reference types can't be shared across initilialization

**Comment by dguy** [comment]
@apurvam - we need a `Supplier` for the aggregates as it is supplying the `initialValue` for an aggregate. The `initialValue` should be a `new` instance/value each time the initializer is called otherwise we end up in a scenario where mulitple aggregations are sharing the same reference type and ending up with incorrect results, i.e., if it was an array `Object[2]{null, null}`, it is called once and then an aggregation runs and updates the array to `Object {0, null}` - then the initializer runs again (this will happen mulitple times for window operations) - the initializer would return `Object{0, null}` which is incorrect

**Pull request merged**